core/runtime/src/main/java/io/quarkus/runtime/ExecutorTemplate.java

```diff
@@ -54,34 +54,41 @@ public Executor setupRunTime(ShutdownContext shutdownContext, int defaultCoreSiz
                 .setThreadFactory(JBossExecutors.resettingThreadFactory(threadFactory));
         final int cpus = ProcessorInfo.availableProcessors();
         // run time config variables
-        builder.setCorePoolSize(getIntConfigVal(CORE_POOL_SIZE, defaultCoreSize == -1 ? 4 * cpus : defaultCoreSize));
-        builder.setMaximumPoolSize(getIntConfigVal(MAX_POOL_SIZE, defaultMaxSize == -1 ? 10 * cpus : defaultMaxSize));
-        builder.setMaximumQueueSize(getIntConfigVal(QUEUE_SIZE, defaultQueueSize));
-        builder.setGrowthResistance(getFloatConfigVal(GROWTH_RESISTANCE, defaultGrowthResistance));
-        builder.setKeepAliveTime(getIntConfigVal("executor.keep-alive-millis", defaultKeepAliveMillis), TimeUnit.MILLISECONDS);
+        builder.setCorePoolSize(threadsConfig.coreThreads);
+        builder.setMaximumPoolSize(threadsConfig.maxThreads <= 0 ? 8 * cpus : threadsConfig.maxThreads);
+        builder.setMaximumQueueSize(threadsConfig.queueSize);
+        builder.setGrowthResistance(threadsConfig.growthResistance);
+        builder.setKeepAliveTime(threadsConfig.keepAliveTime, TimeUnit.MILLISECONDS);
         final EnhancedQueueExecutor executor = builder.build();
-        shutdownContext.addShutdownTask(new Runnable() {
+
+        Runnable shutdownTask = new Runnable() {
             @Override
             public void run() {
                 executor.shutdown();
                 for (;;)
                     try {
-                        executor.awaitTermination(Long.MAX_VALUE, TimeUnit.HOURS);
+                        if (!executor.awaitTermination(threadsConfig.shutdownTimeout, TimeUnit.MILLISECONDS)) {
+                            List<Runnable> tasks = executor.shutdownNow();
+                            for (Runnable i : tasks) {
+                                //for any pending tasks we just spawn threads to run them on a best effort basis
+                                //these threads are daemon threads so if everything shuts down they will not keep the
+                                //JVM alive
```

Comment:
This will almost certainly result in tasks not running; is this to avoid a deadlock or something?

Reply:
I don't really want it to hang indefinitely if something is wrong. If we hit the timeout then we know something is probably screwy, so this seems like a 'best effort' attempt.

Reply:
That's a valid concern.  I guess we really ought to do something staged, like after a couple of seconds print a message like "Some tasks are taking a long time to finish", then some time after that (configurable is reasonable I guess) we could print an error giving, like stack traces or something.

I'll make a note to look into this in the future, when things are a little less hectic.